### PR TITLE
Change To be able to download order data from all products removed

### PR DESCRIPTION
**Pull request by mt8**
### Scenario

- We have a store handling a large number of downloadable products. It has rules for removing products that are out of stock.
- Currently, if all products purchased by a customer are deleted, the download button will no longer be displayed on the order details page.
- This is a reasonable default, but it would be useful if downloadable assets could be programmatically 'injected' so that downloads can remain available to customers, even after the corresponding product has been deleted *(if* this is what the merchant/developer wishes to do).

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions.

1. This PR only applies to the 'classic' (non-Block driven) experience. Therefore, to test, please ensure your environment is configured appropriately (activating a theme like Storefront can help).
2. Functionally, this should not alter the default experience: it simply adds hooks that *allow* 3PDs to modify that experience, if they wish.

**Test that the conventional experience is unchanged**

1. Login as a customer, and purchase a downloadable product.
2. Once you get to the *order-received* page, freeze! You will not see the download table if the order status is not complete so, as an admin user, go ahead and complete the order (but keep the original tab open).
3. Refresh the *order-received* page, and you should be able to see the downloads table:

<div align="center">Downloads table (classic 'order received' page)</div>

4. Visit the My Accounts page, open the order you just placed. You should again see the Downloads table here.

**Test the newly added hook**

Add the following snippet to a suitable location, such as `wp-content/mu-plugins/test-pr-41449.php`:

```php
<?php

add_filter( 'woocommerce_order_downloads_table_show_downloads', '__return_false' );
```

Repeat the previous testing steps. However, this time, the downloads table should not render.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### Notes

The testing instructions above are only intended to demonstrate that the default experience (when purchasing downloadable products) is unchanged, and that the new filter can be used to modify the experience.  To simulate the described scenario we'd need something a little more involved. Simplified example, mostly adding in case handy for future reference:

```php
$inserted_downloads = function ( $downloads, $order = null ) {
	return [ [
		'download_url'        => 'https://example.com/download/123.pdf',
		'download_id'         => 'ab12cd34ef56gh78',
		'product_id'          => 123,
		'product_name'        => 'Injected product',
		'product_url'         => false,
		'download_name'       => 'Injected download',
		'order_id'            => $order ? $order->get_id() : 123,
		'order_key'           => $order ? $order->get_order_key() : 456,
		'downloads_remaining' => 100,
		'access_expires'      => time() + YEAR_IN_SECONDS,
		'file'                => [
			'name' => 'Injected file',
			'file' => 'abcdef12345',
		],
	] ];
};

add_filter( 'woocommerce_order_get_downloadable_items', $inserted_downloads, 10, 2 );
add_filter( 'woocommerce_customer_get_downloadable_products', $inserted_downloads, 10, 2 );
add_filter( 'woocommerce_order_downloads_table_show_downloads', '__return_true' );
```

### CODE REVIEW

**Comment by barryhughes** [comment]
Thanks for the detailed explanation, @mt8.

> I'm developing a store that specializes in selling downloadable products.

What you've described does not sound like a very typical use-case and, given this note, I suppose my first question is if we need to add a further hook to the core plugin, or if you can simply implement one or more [template overrides](https://woo.com/document/template-structure/) within the store you are developing?

If that is not possible, I think we should take a moment to think about the best location for a new filter hook, as modifying the result of a call to `WC_Order::has_downloadable_item()` seems like it could introduce confusion. Could something like the following be preferable instead?

- Extract the [`$show_downloads = $order->has_downloadable_item() && $order->is_download_permitted()`](https://github.com/woocommerce/woocommerce/blob/8.3.0/plugins/woocommerce/templates/order/order-details.php#L30) logic from the `order-details.php` template.
- Move it inside the [`woocommerce_order_details_table()`](https://github.com/woocommerce/woocommerce/blob/8.3.0/plugins/woocommerce/includes/wc-template-functions.php#L2716) function, making it filterable and passing the resulting variable to the template. 

The key difference is that this makes it possible to filter the decision to *show downloads*, vs making it possible to filter logic that determines if downloadable items exist in the context of an order.

**Comment by mt8** [comment]
@barryhughes 

Thanks for your reviewing!

> if you can simply implement one or more [template overrides](https://woo.com/document/template-structure/) within the store you are developing?

Yes, as you pointed out, I used template overrides to get around this problem.

Specifically, the steps are as follows.

1. Copy `/templates/order/order-detail.php` from WooCommerce into the plugin

2. Change the handling of the variable `$show_downloads` as follows.

Change before:

````
$show_downloads = $order->has_downloadable_item() && $order->is_download_permitted();
````

After change:

````
$show_downloads = apply_filters( 'woocommerce_order_has_downloadable_item', $order->has_downloadable_item() ) && $order->is_download_permitted();
````

3. Use the `woocommerce_locate_template` filter to load `/templates/order/order-detail.php` with the following priority

1: Inside the activated theme
2: Things inside custom plugins

By responding in this way, the objective was achieved.

However, this method is not perfect.
This is because the plugin I am developing is expected to be used by an unspecified number of people.

It is fine if `/templates/order/order-detail.php` does not exist in the theme used by the plugin user, but if it exists, it is necessary to ask the user to install a filter. That was difficult for many store owners.

> If that is not possible, I think we should take a moment to think about the best location for a new filter hook, as modifying the result of a call to WC_Order::has_downloadable_item() seems like it could introduce confusion. Could something like the following be preferable instead?

Yes, I think that method is very good.
My goal is not to add a filter, but to be able to download order data from all products removed.

thank you

**Comment by barryhughes** [comment]
> This is because the plugin I am developing is expected to be used by an unspecified number of people.

OK, I just wanted to check in case template overrides were a suitable solution. Sounds like they will not be :+1:

> My goal is not to add a filter, but to be able to download order data from all products removed.

I understand—but, of course, a filter is needed to achieve this goal and what I was discussing in my last comment was the placement of this filter.

- Your proposal is to add it to [`WC_Order::has_downloadable_item()`](https://github.com/woocommerce/woocommerce/blob/3baa6a891e985c88ea828987ed00281c55daf2ee/plugins/woocommerce/includes/class-wc-order.php#L1668) and I'm suggesting that is risky, because it means the method may return unexpected results from the perspective of other plugins.
- It looks like the reason you are doing this, ultimately, is to manipulate the [`$show_downloads`](https://github.com/woocommerce/woocommerce/blob/8.3.0/plugins/woocommerce/templates/order/order-details.php#L30) variable.
- The alternative approach I outlined seems safer, as we would be manipulating the decision to *show* downloads, rather than information about whether an order *contains* downloads.

**Comment by mt8** [comment]
@barryhughes 

Thanks!

> The alternative approach I outlined seems safer, as we would be manipulating the decision to show downloads, rather than information about whether an order contains downloads.

I perfectly agree with your approach.

Shall I edit the code in this pull request?

**Comment by barryhughes** [comment]
Yes please, and just re-request my review when ready :+1:

**Comment by mt8** [comment]
@barryhughes 

I reverted my initial code changes and incorporated code changes inspired by your approach. 
I would appreciate it if you could re-review it. 

thank you

https://github.com/woocommerce/woocommerce/pull/41449/commits/b337e41bd592ac81bc92e6822a0881a305ff69be

**Comment by barryhughes** [comment]
- Testing instructions updated.
- Small tweaks made to satisfy coding standards.

**Comment by barryhughes** [comment]
Thanks for the updates, @mt8:

- As you may be aware, there are basically two separate storefront experiences supported by WooCommerce: 'classic', and Blocks-driven. Your change will only impact the classic experience (and I'm noting this mostly as an FYI, so you aren't caught off-guard at a later point).
- I notice that this new filter does not impact the **My Account ▸ Downloads** page. I don't think this impacts your specific use-case, but it means there is an inconsistency for other developers who may wish to use this new hook.

That last point means we may need a further change. One possibility is that, instead of loading the `order-downloads` template directly from within the `order-details` template, we instead call `woocommerce_order_downloads_table()` and place the new filter inside that function.

**Comment by mt8** [comment]
@barryhughes 

I understood the problem and the contradiction.

Admittedly, in my specific use case, I had deleted the My Account > Downloads page, so it wasn't a problem.

> we instead call woocommerce_order_downloads_table() and place the new filter inside that function.

Yes, I think this is good too.

However, it seems difficult to move this filter inside `woocommerce_order_downloads_table()`. This is because `woocommerce_order_downloads_table()` does not have any arguments related to orders.

Also, note that the parameter `show_title` is set when reading `order/order-downloads.php` from `order/order-details.php`.

I understand that My Count > Downloads displays a list of downloadable products that the customer has been given permission to download.

How about adding this filter to `wc_get_customer_available_downloads()`?

```
diff --git a/plugins/woocommerce/includes/wc-user-functions.php b/plugins/woocommerce/includes/wc-user-functions.php
index cad8096721..aa64e7ec33 100644
--- a/plugins/woocommerce/includes/wc-user-functions.php
+++ b/plugins/woocommerce/includes/wc-user-functions.php
@@ -633,6 +633,10 @@ function wc_get_customer_available_downloads( $customer_id ) {
 				continue;
 			}
 
+			if ( ! apply_filters( 'woocommerce_order_downloads_table_show_downloads', true, $order ) ) {
+				continue;
+			}
+
 			$product_id = intval( $result->product_id );
 
 			if ( ! $_product || $_product->get_id() !== $product_id ) {
```

**Comment by barryhughes** [comment]
Sorry for the delay... I may be misunderstanding, but I'm a little unsure what advantage that provides over existing filter `woocommerce_customer_available_downloads`?

> This is because woocommerce_order_downloads_table() does not have any arguments related to orders.

True, though you can access the order ID for each downloadable item via the elements inside the `$downloads` array (also true for the other hook I mentioned).

**Comment by mt8** [comment]
@barryhughes 

> True, though you can access the order ID for each downloadable item via the elements inside the $downloads array (also true for the other hook I mentioned).

thank you. I'll try checking this.

**Comment by mt8** [comment]
@barryhughes

I made some changes locally and did some testing.

> That last point means we may need a further change. One possibility is that, instead of loading the order-downloads template directly from within the order-details template, we instead call woocommerce_order_downloads_table() and place the new filter inside that function.

Regarding this, I was able to empty the My Account > Downloads display by following the steps below.

- In `order-details.php`, execute `woocommerce_order_downloads_table()` without loading `order-downloads.php`.
- Add a new temporary filter `woocommerce_order_downloads_table_enabled` to `woocommerce_order_downloads_table()` to determine whether to load `order-downloads.php`
- Write the following in `mu-plugin/test.php`

````
add_filter( 'woocommerce_order_downloads_table_show_downloads', '__return_false' );
add_filter( 'woocommerce_order_downloads_table_enabled', '__return_false' );
````

However, this is what it looks like at that time.


The message "No downloads available yet." is not displayed.

I originally expected it to look like this:


I checked the code further and finally realized that I could just return `empty` for `woocommerce_customer_get_downloadable_products` inside `WC()->customer->get_downloadable_products()` in `downloads.php`. I did.

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/myaccount/downloads.php#L24
https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-customer.php#L233

In other words, write the following two statements in `mu-plugin/test.php`

````
// new filter
add_filter( 'woocommerce_order_downloads_table_show_downloads', '__return_false' );

// inside of `WC()->customer->get_downloadable_products()`
add_filter( 'woocommerce_customer_get_downloadable_products', '__return_empty_array' );
````

By doing so, It can control the download display in the following two places:

- My Account > Orders > Order Details Page > Download
- My Account > Downloads

In conclusion, since My Account > Downloads can already be controlled using existing filter, this pull request would be good if it would be possible to control the display of download items in My Account > Orders > Order Details.

Thanks

**Comment by barryhughes** [comment]
Thanks, this is looking pretty good!

**Review by barryhughes — APPROVED** [review]
Once checks pass, we can merge :+1:

**Comment by mt8** [comment]
@barryhughes 
Thank you so much 😉

**Comment by barryhughes** [comment]
You're welcome—thanks for the contribution!

I'll re-run the failing E2E tests. The failure looks like it is completely unrelated:

```
Testing stopped early after 15 maximum allowed failures.
  1 failed
    [Chrome] › merchant/products/add-variable-product/update-variations.spec.js:407:2 › Update variations › can manage stock levels 
```

Less sure why the template version check is failing, but I'll check in on that one and make further adjustments if required.

**Comment by barryhughes** [comment]
Rebased to hopefully solve the failing version tag check.
